### PR TITLE
feat(training): add cache=ram|disk image cache (Ultralytics parity, #247)

### PR DESCRIPTION
**Pull request by TreschQ**
<h2>Summary</h2>
<p>Closes part of #247 (Ultralytics API parity audit). Adds the <code>cache</code>
parameter to <code>TrainConfig</code> and wires it through <code>BaseTrainer</code> into
<code>YOLODataset</code> / <code>COCODataset</code>, mirroring Ultralytics' <code>BaseDataset</code>
caching contract (<code>cache=False | True | "ram" | "disk"</code>).</p>
<p>Until now, <code>model.train(data=..., cache=...)</code> emitted
<code>UserWarning: Unknown training config keys (ignored): ['cache']</code>
and silently did nothing. Users hitting an IO-bound dataloader (slow
storage, HD source images, large batch on a fast GPU) had no way to
opt into the standard fix.</p>
<h2>What changed</h2>
<ul>
<li><strong><code>libreyolo/training/config.py</code></strong> — new <code>cache: bool | str | None = False</code>
field on <code>TrainConfig</code> (inherited by every family config: YOLO9, RF-DETR,
D-FINE, …). Documented values: <code>False</code> / <code>None</code> / <code>True</code> / <code>"ram"</code> / <code>"disk"</code>.</li>
<li><strong><code>libreyolo/data/dataset.py</code></strong> — shared helpers
(<code>_normalize_cache</code>, <code>_check_cache_ram</code>, <code>_check_cache_disk</code>,
<code>_cache_image_to_disk</code>, <code>_available_memory_bytes</code>) plus
<code>_init_image_cache</code> / <code>_populate_image_cache</code> methods on both
<code>YOLODataset</code> and <code>COCODataset</code>. <code>load_image</code> reads <code>.npy</code> when present;
<code>load_resized_img</code> returns the RAM-cached array when present. Caching
pass uses <code>ThreadPool(NUM_THREADS)</code> + <code>tqdm</code> with Ultralytics-style
<code>Caching images (X.XGB {RAM|Disk})</code> description.</li>
<li><strong><code>libreyolo/training/trainer.py</code></strong> — <code>_setup_data</code> forwards
<code>self.config.cache</code> to every <code>YOLODataset</code> / <code>COCODataset</code> construction
site.</li>
<li><strong><code>tests/unit/test_image_cache.py</code></strong> — 22 new unit tests.</li>
</ul>
<h2>Semantics (matches Ultralytics)</h2>
<ul>
<li><code>cache=False</code> / <code>None</code> (default) — unchanged behaviour.</li>
<li><code>cache=True</code> / <code>"ram"</code> — decode + resize all images once at dataset
construction, store the resulting <code>uint8</code> arrays in <code>self.ims[i]</code>.
<code>load_resized_img(i)</code> returns directly from RAM (zero IO).</li>
<li><code>cache="disk"</code> — write each decoded BGR image as <code>.npy</code> next to the
source. <code>load_image(i)</code> reads <code>.npy</code> (no JPG decode); resize still
happens at runtime.</li>
<li>Pre-flight check with 50% safety margin (sample 30 images, extrapolate,
compare to available RAM / free disk). On failure: log a warning and
fall back to no cache — <strong>never crashes the run</strong>.</li>
</ul>
<h2>No new dependency</h2>
<p><code>psutil</code> is used opportunistically; when missing we fall back to
<code>os.sysconf("SC_PAGE_SIZE") * os.sysconf("SC_AVPHYS_PAGES")</code> (POSIX),
then to <code>SC_PHYS_PAGES</code>, then to 0 (treated as "unknown, skip caching").</p>
<h2>Before / after (synthetic micro-bench)</h2>
<p>Tiny CPU run, 5 train + 2 val 640×640 images, YOLO9-t, 1 epoch:</p>

cache | epoch time | unknown-keys warning
-- | -- | --
False | 1.27 s | no
"ram" | 1.09 s | no
"disk" | 1.15 s | no

<p>The gain is in the noise at this scale (3 batches, decode is not the
bottleneck). Real-world measurement, 4163 train + 1040 val 1080p images,
YOLO9-t, imgsz=320, batch=16, MPS, <code>cache="ram"</code>:</p>
<ul>
<li>Cache fill (one-shot at dataset construction): <strong>~5 s</strong> for 0.72 GB.</li>
<li>Epoch 1 (including the cache fill): 3 min 45 s, loss = 4.3787 (finite).</li>
<li>Subsequent epochs skip JPG decode + resize entirely; the GPU-starvation
pattern described in #247 (e.g. A40 batch=128 @ 12 % GPU usage) is
the canonical target scenario.</li>
</ul>
<h2>Test plan</h2>
<ul>
<li>[x] <code>pytest tests/unit/test_image_cache.py</code> — 22/22 pass.</li>
<li>[x] No regression on the existing dataset / train-cfg / yolo9 smoke
tests (<code>pytest tests/unit/test_dataset_loading.py     tests/unit/test_train_cfg.py tests/unit/test_yolo9_e2e_trainer_smoke.py     tests/unit/test_data_utils.py tests/unit/test_train_callbacks.py</code>
→ 70/70 pass, including the 22 new ones).</li>
<li>[x] End-to-end smoke: <code>model.train(cache="ram")</code> on a real 4163-image
YOLOv9-t fine-tune — no warning emitted, cache filled, epoch 1
converges to a finite loss.</li>
<li>[ ] Reviewer: ideally one A40-class GPU benchmark vs <code>cache=False</code> on a
real dataset to confirm the epoch-time delta predicted in #247.</li>
</ul>
<h2>Out of scope</h2>
<ul>
<li>The same <code>cache</code> knob in <code>ValidationConfig</code>. Validators already iterate
the dataset linearly and benefit less; happy to do it as a follow-up.</li>
<li>Segmentation <code>load_image</code> (unresized) path: when <code>cache="ram"</code> is used
with a segmentation trainer (<code>wants_unresized_image=True</code>), the RAM
cache populates but is not consulted because the resized contract
differs. <code>cache="disk"</code> works for both. Same trade-off as Ultralytics.</li>
</ul></body></html>

### CODE REVIEW

**Comment by TreschQ** [comment]
I am currently benchmarking cache vs no cache on custom fine tuning on A2000 RTX ADA with ~3600 train images on yolov9t (200 epochs)

**Comment by EHxuban11** [comment]
Hello @TreschQ , thank you for considering contributing to LibreYOLO. Please, open an Issue first and let's agree on the problem and the approach to fix it before we commit to a code solution. 

https://github.com/LibreYOLO/libreyolo/blob/dev/CONTRIBUTING.md
